Add IndexTTS

senstella opened the pull request:
## Context
Adds [IndexTTS](https://github.com/index-tts/index-tts)

## Description
Implements `Conformer` and `Perceiver`. The main model is based on GPT2 architecture, it will leverge `mlx-lm`'s GPT-2 implementation.

## Additional information
This is not done yet. For now, it loads weights correctly. I need to write proper forward pass later.

<!-- Optional: Add a checklist for contributors -->
## Checklist
- [X] Loads weight
- [X] Do forward pass correctly
- [X] Decode the generated mel via BigVGAN
- [X] Tests added/updated

senstella commented:
@Blaizzy I just noticed IndexTTS uses non-standard BigVGAN, which incoperates conditioning inside of BigVGAN using an ECAPA-TDNN model, such as:

```diff
        for i in range(self.num_upsamples):
            # Upsampling
            for i_up in range(len(self.ups[i])):
                x = self.ups[i][i_up](x)

+           if self.cond_in_each_up_layer:
+               x = x + self.conds[i](speaker_embedding)
```

Since it's not the standard BigVGAN implementation, I'm thinking of just implementing it in `tts/model/indextts/bigvgan.py`, or perhaps I should implement it somewhere like `codec/model/bigvgan_indextts` since it's still a codec?

What do you think would be more reasonable?

Blaizzy commented:
Awesome work @senstella!

>@Blaizzy I just noticed IndexTTS uses non-standard BigVGAN, which incoperates conditioning inside of BigVGAN using an ECAPA-TDNN model,

I agree with @lucasnewman 100% here. It's the approach we usually use.

Rule of thumb:
Start with having the custom logic with the model (i.e., IndexTTS), and if we see a similar pattern on the next model we port with BigVGAN codec, then we migrate it there.

Great example that demostrates this is SparkTTS, it uses ECAPA-TDNN and some custom modules with vocos Backbone and DAC layers.

Read more here: https://github.com/Blaizzy/mlx-audio/tree/main/mlx_audio/tts/models/spark/modules/encoder_decoder

Blaizzy commented:
Thank you very much @senstella! 

I will review and merge it by Sunday 🚀

lucasnewman commented on mlx_audio/codec/models/bigvgan/bigvgan.py:
```diff
@@ -116,3 +126,41 @@ def __call__(self, x: mx.array) -> mx.array:  # (batch, num_mels, seq)
+
+            if key.endswith("weight_v"):
+                v_norm = (
+                    normalize_weight(value, 0)
```
This likely works fine, but I wonder if this should just be a WNConv1D instance instead of handling the normalization in weight conversion. Then it match the torch implementation and make it possible to fine-tune the model if we ever decide to want that. There are some examples of it in other areas of the codebase so it might just be a drop-in thing.

senstella replied:
Sounds awesome, and that's more faithful to the original implementation! I'll switch to `WNConv1D` and`WNConvTranspose1d` and re-upload the weights using the new conversion format.

lucasnewman commented on mlx_audio/tts/models/indextts/indextts.py:
```diff
@@ -0,0 +1,403 @@
+
+        mel_position = 0
+
+        for _ in range(max_tokens):
```
Very minor: might be nice to use tqdm.trange here when verbose = True for the progress display.

lucasnewman reviewed (approved):
This looks really well done, thank you!

lucasnewman commented:
I noticed the WER in English seems quite a bit higher than they advertise in their results -- is that the case with the torch model as well? It does sound great when it gets the words right and the voice matching is impressive.

chigkim commented:
When you generate via python script, isn't ref_audio supposed to accept file name, not mlx array? I think other models accept file names if I'm not mistaken?

chigkim commented:
Oh, interesting, it looks like most models accept mlx.array, but Spark accepts Path, and Outetts accepts str?

senstella commented on mlx_audio/tts/models/indextts/conformer.py:
```diff
@@ -0,0 +1,240 @@
+    def __init__(self, args: ConformerArgs):
+        super().__init__()
+        conv_layers = self.CONV_LAYERS[args.input_layer]
+        # mask_patterns = self.CONV_MASKS[args.input_layer]
```
I added the masking in 58d7e3bd01693f55a126a77c1f8b14fc61ae196d for the case when conformer wants to encode different length of samples!

chigkim commented:
I was wrong generate function from other models also accepts mx.array. Sorry about that.

senstella commented:
I've added a `conv.py` file with the `WNConv1d` and `WNConvTranspose1d` classes from `descript/nn/layers.py`, also I've updated the weights on HF. Should be all set now!

Blaizzy commented:
> Note that the test failure is related to https://github.com/Blaizzy/mlx-audio/issues/191, which is caused by mlx-lm's API refactoring.

@senstella fixed in #194

Blaizzy commented:
@senstella the biganv tests are failing, could you look at them?

senstella commented:
Test should pass now, the problem was `WNConvTranspose1d` calls `mx.conv_transpose1d` with parameters
```py
y = mx.conv_transpose1d(
    x, weight, self.stride, self.padding, self.dilation, self.groups
)
```
while `mx.conv_transpose1d` accepts:
```py
(input, weight, stride, padding, dilation, output_padding, groups)
```
Therefore, the default value of group(1) slid into output_padding, which messed up the shaping.

Blaizzy commented:
Awesome, thanks for the fix!

One more thing I just noticed. The converted model has the codec weights embedded within it.

I typically keep codec weights separate from the main model, which allows us to quantize only the model weights while leaving the codec unquantized. This approach helps preserve output quality since codec degradation can significantly impact audio fidelity.

There's limited data on my approach just intuition so I could be off base here. Would you mind testing the quantized versions (i.e., 4bit or 6bit) to confirm the audio output still sounds acceptable?

senstella commented:
I think [MLX only quantizes modules that implement `to_quantized(group_size, bits)`](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.quantize.html#mlx.core.quantize). However, BigVGAN codec doesn't contain any `Linear` nor `Embedding`. And [`Conv1d` layers doesn't have `to_quantized` method](https://ml-explore.github.io/mlx/build/html/python/nn/_autosummary/mlx.nn.Conv1d.html), so in theory the codec wouldn't be quantized.

Here's some quantized model's generations and full model generation (group_size: 64): [4 bits](https://files.catbox.moe/rrdusx.wav) / [6 bits](https://files.catbox.moe/rrdusx.wav) / [fp16](https://files.catbox.moe/8cfsyo.wav)

I also just realized I forgot to implement the text normalizer. It seems like English can be mostly mimiced in Python level easily. (Just replace some characters(: -> ,) and verbalize the number(4 -> four)) However, I'm not quite sure about [Chinese](https://github.com/index-tts/index-tts/blob/10d557a15e0bc234389a2900b7147c4c8a94fe3b/indextts/utils/front.py#L121C3-L126C46). The package uses `wetext` which downloads the normalizer from ModelScope. (HuggingFace of China, I believe) Should I include the `wetext` package in the dependency, or just do the essential normalization in Python? I personally think minimal dependency would be favorable, but I'm not sure basic Python normalization would be sufficient for Chinese texts.

senstella commented:
I just included the original `wetext` normalizer, but please let me know if you prefer otherwise!

Blaizzy commented:
> I think [MLX only quantizes modules that implement to_quantized(group_size, bits)](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.quantize.html#mlx.core.quantize). However, BigVGAN codec doesn't contain any Linear nor Embedding. And [Conv1d layers doesn't have to_quantized method](https://ml-explore.github.io/mlx/build/html/python/nn/_autosummary/mlx.nn.Conv1d.html), so in theory the codec wouldn't be quantized.

Awesome, 

I heard the audio, sounds good to me ✅

Thansk for clarifying and testing because this is true for vocos, snac and mimi as they have `nn.Linear` layers.

Blaizzy commented:
> I just included the original wetext normalizer, but please let me know if you prefer otherwise!

Great, I like the normalizer!

However, I wonder if it's possible to normalize the text without adding extra dependencies? This will keep us light and make the swift port easier as well.

Blaizzy reviewed (approved):
LGTM!

bytefer commented:
@senstella Great to see mlx-audio supporting IndexTTS! Thank you so much for your efforts. Since [IndexTTS 2.0](https://huggingface.co/IndexTeam/IndexTTS-2) has been released, I'm wondering if you have any plans to support IndexTTS 2.0 as well.

senstella commented:
> > > @senstella Great to see mlx-audio supporting IndexTTS! Thank you so much for your efforts. Since [IndexTTS 2.0](https://huggingface.co/IndexTeam/IndexTTS-2) has been released, I'm wondering if you have any plans to support IndexTTS 2.0 as well.
> > 
> > 
> > Thanks for letting me know! I'll take a look at the paper try implementing them when I get some time. Was a little busy these days, sorry!
> 
> I am glad to receive your reply. Thanks to your help, I have successfully run IndexTTS-1.5 on my m1 laptop, and found a problem in using it, IndexTTS does not need to use the `ref_text` parameter, SparkTTS does. This results in running IndexTTS without the `ref_text` parameter set, which automatically downloads “mlx-community/whisper-large-v3-turbo” to convert the audio to text.
> 
> ```shell
> Ref_text not found. Transcribing ref_audio...
> Fetching 4 files: 100%|█████████████████████████████████████████████| 4/4 [00:00<00:00, 44034.69it/s]
> Ref_text  So, just to clarify, we've had 19 to 20 year olds on our podcast so far, and if you don't mind me asking, how old are you?
> ```
> 
> [mlx_audio/tts/generate.py](https://github.com/Blaizzy/mlx-audio/blob/3edbfb767d0ae36da497f420d170acb09049875f/mlx_audio/tts/generate.py#L262)
> 
> ```python
>            if not ref_text:
>                print("Ref_text not found. Transcribing ref_audio...")
>                from mlx_audio.stt.models.whisper import Model as Whisper
> 
>                stt_model = Whisper.from_pretrained(path_or_hf_repo=stt_model)
>                ref_text = stt_model.generate(ref_audio).text
>                print("Ref_text", ref_text)
> 
>                # clear memory
>                del stt_model
>                mx.clear_cache()
> ```
> 
> Is it possible to avoid the above issue with the following code, which I have tested locally is working fine. Help to confirm this issue when you have time. Thanks and looking forward to your reply.
> 
> ```python
>             if not ref_text:
>                 import inspect
> 
>                 if "ref_text" in inspect.signature(model.generate).parameters:
>                     print("Ref_text not found. Transcribing ref_audio...")
>                     from mlx_audio.stt.models.whisper import Model as Whisper
> 
>                     stt_model = Whisper.from_pretrained(path_or_hf_repo=stt_model)
>                     ref_text = stt_model.generate(ref_audio).text
>                     print("Ref_text", ref_text)
> 
>                     # clear memory
>                     del stt_model
>                     mx.clear_cache()
> ```

Yeah, current implementation transcribes the `ref_text` no matter what model demands. Using `inspect` to check the signature looks good to me! You might want to create a new pull request to the repository and get it merged!
